blocks/certifier.go

```diff
@@ -0,0 +1,404 @@
+func WithCertConfig(cfg CertConfig) CertifierOpt {
+	return func(c *Certifier) {
+		c.cfg = cfg
+	}
+}
+
+// WithCertifierLogger defines logger for Certifier.
+func WithCertifierLogger(logger log.Log) CertifierOpt {
+	return func(c *Certifier) {
+		c.logger = logger
+	}
+}
+
+type certInfo struct {
+	deadline         time.Time
+	done             bool
+	totalEligibility uint16
+	signatures       []types.CertifyMessage
+}
+
+type certState int
+
+const (
+	unknown certState = iota
+	pending
+	certified
+	expired
+)
+
+// Certifier collects enough CertifyMessage for a given hare output and generate certificate.
+type Certifier struct {
+	logger log.Log
+	cfg    CertConfig
+	once   sync.Once
+	eg     errgroup.Group
+	ctx    context.Context
+	cancel func()
+
+	db         *sql.Database
+	oracle     hare.Rolacle
+	nodeID     types.NodeID
+	signer     *signing.EdSigner
+	publisher  pubsub.Publisher
+	layerClock layerClock
+
+	mu       sync.Mutex
+	certMsgs map[types.LayerID]map[types.BlockID]*certInfo
+}
+
+// NewCertifier creates new block certifier.
+func NewCertifier(
+	db *sql.Database, o hare.Rolacle, n types.NodeID, s *signing.EdSigner, p pubsub.Publisher, lc layerClock,
+	opts ...CertifierOpt,
+) *Certifier {
+	c := &Certifier{
+		logger:     log.NewNop(),
+		cfg:        defaultCertConfig(),
+		ctx:        context.Background(),
+		db:         db,
+		oracle:     o,
+		nodeID:     n,
+		signer:     s,
+		publisher:  p,
+		layerClock: lc,
+		certMsgs:   make(map[types.LayerID]map[types.BlockID]*certInfo),
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	c.ctx, c.cancel = context.WithCancel(c.ctx)
+	return c
+}
+
+// Start starts the background goroutine for periodic pruning.
+func (c *Certifier) Start() {
+	c.once.Do(func() {
+		c.eg.Go(func() error {
+			return c.run()
+		})
+	})
+}
+
+// Stop stops the outstanding goroutines.
+func (c *Certifier) Stop() {
+	c.cancel()
+	err := c.eg.Wait()
+	if err != nil && !errors.Is(err, context.Canceled) {
+		c.logger.With().Error("blockGen task failure", log.Err(err))
+	}
+}
+
+func (c *Certifier) isShuttingDown() bool {
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *Certifier) run() error {
+	for layer := c.layerClock.GetCurrentLayer(); ; layer = layer.Add(1) {
+		select {
+		case <-c.layerClock.AwaitLayer(layer):
+			_ = c.prune()
+		case <-c.ctx.Done():
+			return fmt.Errorf("context done: %w", c.ctx.Err())
+		}
+	}
+}
+
+// RegisterDeadline register the deadline to stop waiting for enough signatures to generate
+// the certificate.
+func (c *Certifier) RegisterDeadline(lid types.LayerID, bid types.BlockID, now time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, ok := c.certMsgs[lid]; !ok {
+		c.certMsgs[lid] = make(map[types.BlockID]*certInfo)
+	}
+
+	c.certMsgs[lid][bid] = &certInfo{
+		deadline:   now.Add(time.Duration(c.cfg.SignatureWaitDuration) * time.Second),
+		signatures: make([]types.CertifyMessage, 0, c.cfg.CertifyThreshold),
```

Comment:
not sure what you mean. do you mean it makes no difference what the capacity is?

Reply:
i meant that we don't number of signatures to be equal to this threshold. we expect eligibility count to be equal to threshold

Reply:
right. but i still need to initialize it to some capacity, and given that we don't use fraction for eligibility count, this is a safe number to use. or the committee size.

Reply:
Yeah, the CertifyThreshold is the maximum it can be. (i.e. everybody's signer's weight is 1)

Reply:
Maybe a comment is required to explain the choice?

Reply:
why do you think this matters so much? 

for me it really doesn't matter. whether it's CertifyThreshold or CommitteeSize, they are O(1) difference. we could also just do make([]types.CertifyMessage, 0) and call it a day...

sure. next time i have something meaningful to update, i'll add a comment.

Reply:
I agree it doesn't matter so much. But I'm thinking about when we have contributors from the wild they'll want to know why.

Reply:
right. so if i change the code to `make([]types.CertifyMessage, 0)`, there will be no confusion because the array will just grow organically as we receive message. no comments needed.

Reply:
I guess so. seems like a reasonable enough sacrifice in the name of understandability.

Reply:
changed back to CommitteeSize because i love to come full circle.

Reply:
omg stop whyyyy, I'm changing my review to needs work

Reply:
exactly why do you think this need work? this is the maximum capacity that can be possible. as such, no comments is required. adding unnecessary comments will only serve to confuse.

you yourself said
> If it was enterprise software, I would have left it at CommitteeSize when I read it the first time.

please unblock. this is not cool. @jonZlotnik

Reply:
I would have left it at CommitteeSize because it doesn't matter in enterprise software. You have your colleagues to explain things to you so I would have just ignored that it was too big an array for what it was. `CommitteeSize` makes it more confusing than `CertifyThreshold`

The maximum possible capacity that it would grow to is `CertifyThreshold` because if we assume every committee member has one eligibility, your implementation collects signatures until CertifyThreshold number of signatures is reached, and then builds/saves the certificate. It drops any further signatures (as it should):
https://github.com/spacemeshos/go-spacemesh/blob/a5269257c54ee59fc082f37b9b2ee2f6431d1a33/blocks/certifier.go#L340-L351

If any committee member has any more than 1 eligibility, the number of signatures required to reach CertifyThreshold decreases.

Reply:
I'm okay with either:
 `make([]types.CertifyMessage, 0, c.cfg.CertifyThreshold)`
or 
`make([]types.CertifyMessage, 0)`

Reply:
my rationale is, committee size is the max size it can grow for any implementation. the current impl, the max is the threshold yes. but say we decided to wait for ALL msgs before we certify? and because it matters so little (800 vs 401), i decided to go for the future-safe one.

Reply:
okidoki